### DISCUSSION TITLE
SimpleMeshLayer with mesh from a PNG does not work with texture.

### DISCUSSION THREAD

**Discussion by nilscb** [visgl/deck.gl#6421] [Q&A] [answered]
Hi

Maybe I can post a question here that has been bugging me a lot.
I am having some trouble displaying a SimpleMeshLayer. It gets its mesh from an image in PNG format 
and has an image as texture. Without the texture it displays fine both as a wireframe and fully colored. 
But with texture it seems it get colored all white and hence invisible against a white background.
Sometimes I made it work by adding another layer using the same texture in  front of it. But it is not consistent.
Does anyone have any idea what is wrong with the code? It could be bug in DeckGl or more likely a mistake on my part
 as I am not very fluent in DeckGl.

```
    const simple_mesh_layer = new SimpleMeshLayer({
      id: 'simple-mesh-layer'
      , data: [{"position": [-512, -512, 0], "angle": 0,  "color": [0, 255, 0] }]
       //, texture: "colors.png"   // the texture makes the mesh get all colored white  !?
      , mesh: "terrain.png"
      , loaders: [TerrainLoader]
      , loadOptions: {
        terrain: {
          bounds: [0, 0, 1024, 1024],
        }
      }
      , getPosition: d => d.position
      , getColor: d => d.color
      , getOrientation: d => [0, d.angle, 0]
      //, wireframe: true
    });
```

Ps A TerrainLayer works just fine. Like this:

```
    const simple_mesh_layer = new TerrainLayer({
      id: 'simple-terrain-layer'
      , texture: "colors.png"
      , elevationData: "terrain.png"
      , bounds: [0, 0, 1024, 1024]
    });
```
	 

Regards
Nils Christian Bonnevie

Attached are the two pictures used 

![terrain](https://user-images.githubusercontent.com/136738/143271660-1315e9cb-656d-43e7-b147-dbefdbc5575e.png)
<img width="128" alt="colors" src="https://user-images.githubusercontent.com/136738/143271695-754f8d07-fcc0-4e05-9866-e768f6962753.png">

**Comment by kylebarron** [comment]
I don't see why you wouldn't want to use the `TerrainLayer` directly; that's its purpose.

If you really want to use a `SimpleMeshLayer`, the `TerrainLayer` is a subclass of it, so I'd recommend reading its source: https://github.com/visgl/deck.gl/blob/master/modules/geo-layers/src/terrain-layer/terrain-layer.js

**Comment by nilscb** [comment]
Yes I should probably have explained why I want to use SimpleMeshLayer. The plan was to extend it and modify the fragment shader so that it would be possible to use the texture pixels not to directly shade 
the surface but to use as a lookup in a provided color map. But the texture pixels came out all white just like in the "of the self usage" described in the code fragment I gave. That's why I suspect a defect in the SimpleMeshLayer. But then again the terrain layer (a composite layer) uses SimpleMeshLayer as a sublayer and that works so obviously there is a way to make set up SimpleMeshLayer correctly. I am still a bit puzzled though that my code example did not work. But thanks for the link I will try to understand it. Maybe make my own composite layer using an exetended SimpleMeshLayer
could be a solution.

**Comment by Pessimistress** [comment, marked as answer]
Your `SimpleMeshLayer` parses the texture as terrain instead of as an image. The `TerrainLoader` that you supply has higher priority than the default loaders (`ImageLoader` in this case) and it intercepts all png data sources.

You need to load the terrain manually like this:

```js
import {load} from '@loaders.gl/core';
import {TerrainLoader} from '@loaders.gl/terrain';

new SimpleMeshLayer({
  texture: '/color.png',
  mesh: load('/terrain.png', TerrainLoader, {
    // options
  },
  ...
});
```

**Comment by nilscb** [comment]
That worked! This saved me a lot of headache I don't thing I would be able to figure this out myself. I'm happy I found this place to ask questions. Thank you very much!

Nils Christian
